(3.1.1-snapshot) db error on upgrade due to webhooks

**Describe the bug**

Existing auth tokens in DB cause error in at startup with webhooks feature:

>Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: NULL not allowed for column "TYPE"; SQL statement:
alter table auth_token add column type varchar(255) not null [23502-199]

and

> Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Column "TYPE" not found; SQL statement:
create index IDX_TYPE on auth_token (type) [42122-199]

**My Rundeck detail**

* Rundeck version: 3.1.1-snapshot

**To Reproduce**
Steps to reproduce the behavior:
1. Create auth tokens in 3.1.0/3.1.0-snapshot environment
2. Upgrade to 3.1.1-snapshot
3. Start rundeck

**Expected behavior**
no errors

**Additional context**

`nullable:true` constraint should be added to the `type` field for AuthToken

--- a/rundeckapp/grails-app/controllers/rundeck/controllers/UserController.groovy
+++ b/rundeckapp/grails-app/controllers/rundeck/controllers/UserController.groovy
@@ -131,7 +131,11 @@ class UserController extends ControllerBase{
             if (!tokenAdmin) {
                 eq("creator", u.login)
             }
-            eq("type", AuthTokenType.USER)
+            or {
+                eq("type", AuthTokenType.USER)
+                isNull("type")
+            }
+
         }
 
         int max = (params.max && params.max.isInteger()) ? params.max.toInteger() :
@@ -160,7 +164,10 @@ class UserController extends ControllerBase{
             if (max) {
                 maxResults(max)
             }
-            eq("type", AuthTokenType.USER)
+            or {
+                eq("type", AuthTokenType.USER)
+                isNull("type")
+            }
             order("dateCreated", "desc")
         }
         params.max = max

--- a/rundeckapp/grails-app/controllers/rundeck/interceptors/SetUserInterceptor.groovy
+++ b/rundeckapp/grails-app/controllers/rundeck/interceptors/SetUserInterceptor.groovy
@@ -4,6 +4,7 @@ import com.dtolabs.rundeck.core.authentication.Group
 import com.dtolabs.rundeck.core.authentication.Username
 import com.dtolabs.rundeck.core.authentication.tokens.AuthTokenType
 import grails.plugin.springsecurity.SpringSecurityUtils
+import groovy.transform.PackageScope
 import org.rundeck.web.infosec.AuthorizationRoleSource
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.context.ApplicationContext
@@ -62,8 +63,8 @@ class SetUserInterceptor {
         } else if (request.api_version && !session.user ) {
             //allow authentication token to be used
             def authtoken = params.authtoken? params.authtoken : request.getHeader('X-RunDeck-Auth-Token')
-            AuthTokenType tokenType = controllerName == "webhook" && actionName == "post" ? AuthTokenType.WEBHOOK : AuthTokenType.USER
-            String user = lookupToken(authtoken, servletContext, tokenType)
+            boolean webhookType = controllerName == "webhook" && actionName == "post"
+            String user = lookupToken(authtoken, servletContext, webhookType)
             List<String> roles = lookupTokenRoles(authtoken, servletContext)
 
             if (user){
@@ -157,7 +158,8 @@ class SetUserInterceptor {
      * @param context
      * @return
      */
-    private String lookupToken(String authtoken, ServletContext context, AuthTokenType tokenType) {
+    @PackageScope
+    String lookupToken(String authtoken, ServletContext context, boolean webhookType) {
         if(!authtoken){
             return null
         }
@@ -171,7 +173,20 @@ class SetUserInterceptor {
                 return user
             }
         }
-        def tokenobj = authtoken ? AuthToken.findByTokenAndType(authtoken, tokenType) : null
+
+        AuthToken tokenobj = null
+        if(webhookType) {
+            tokenobj = AuthToken.findByTokenAndType(authtoken,AuthTokenType.WEBHOOK)
+        } else {
+            tokenobj = AuthToken.createCriteria().get {
+                eq("token",authtoken)
+                or {
+                    eq("type", AuthTokenType.USER)
+                    isNull("type")
+                }
+            }
+        }
+
         if (tokenobj) {
             if (tokenobj.tokenIsExpired()) {
                 log.debug("loginCheck token is expired ${tokenobj?.user}, ${tokenobj}");

--- a/rundeckapp/grails-app/domain/rundeck/AuthToken.groovy
+++ b/rundeckapp/grails-app/domain/rundeck/AuthToken.groovy
@@ -30,7 +30,7 @@ class AuthToken implements AuthenticationToken {
     Date expiration
     Date dateCreated
     Date lastUpdated
-    AuthTokenType type
+    AuthTokenType type = AuthTokenType.USER
     static belongsTo = [user:User]
     static transients = ['printableToken','ownerName']
     static constraints = {
@@ -42,6 +42,7 @@ class AuthToken implements AuthenticationToken {
         expiration(nullable: true)
         lastUpdated(nullable: true)
         dateCreated(nullable: true)
+        type(nullable: true)
     }
     static mapping = {
         authRoles type: 'text'

--- a/rundeckapp/src/test/groovy/rundeck/controllers/UserControllerSpec.groovy
+++ b/rundeckapp/src/test/groovy/rundeck/controllers/UserControllerSpec.groovy
@@ -1,10 +1,13 @@
 package rundeck.controllers
 
+import com.dtolabs.rundeck.core.authentication.tokens.AuthTokenType
 import com.dtolabs.rundeck.core.authorization.UserAndRolesAuthContext
 import grails.test.mixin.Mock
 import grails.test.mixin.TestFor
 import grails.test.mixin.TestMixin
 import grails.test.mixin.web.GroovyPageUnitTestMixin
+import grails.testing.gorm.DataTest
+import grails.testing.web.controllers.ControllerUnitTest
 import org.grails.web.servlet.mvc.SynchronizerTokensHolder
 import rundeck.AuthToken
 import rundeck.Execution
@@ -19,10 +22,13 @@ import spock.lang.Unroll
 import javax.servlet.http.HttpServletRequest
 import javax.servlet.http.HttpServletResponse
 
-@TestFor(UserController)
-@Mock([Execution, User, AuthToken])
-@TestMixin(GroovyPageUnitTestMixin)
-class UserControllerSpec extends Specification{
+class UserControllerSpec extends Specification implements ControllerUnitTest<UserController>, DataTest {
+
+    void setupSpec() {
+        mockDomain Execution
+        mockDomain User
+        mockDomain AuthToken
+    }
 
     protected void setupFormTokens(params) {
         def token = SynchronizerTokensHolder.store(session)
@@ -466,4 +472,40 @@ class UserControllerSpec extends Specification{
         response.status==200
     }
 
+    def "profile"() {
+        setup:
+        User user = new User(login: "admin")
+        user.save()
+        createAuthToken(user:user,type:null)
+        createAuthToken(user:user,type: AuthTokenType.USER)
+        createAuthToken(user:user,type: AuthTokenType.WEBHOOK)
+        def authCtx = Mock(UserAndRolesAuthContext)
+        controller.frameworkService = Mock(FrameworkService) {
+            getAuthContextForSubject(_) >> authCtx
+            authorizeApplicationResourceType(_,_,_) >> true
+        }
+        controller.apiService = Stub(ApiService)
+
+        controller.metaClass.unauthorizedResponse = { Object tst, String action, Object name, boolean fg -> false }
+        controller.metaClass.notFoundResponse = { Object tst, String action, Object name, boolean fg -> false }
+
+        when:
+        params.login = "admin"
+        def result = controller.profile()
+
+        then:
+        AuthToken.count() == 3
+        result.tokenTotal == 2
+        result.tokenList.size() == 2
+    }
+
+    private void createAuthToken(params) {
+        AuthToken tk = new AuthToken()
+        tk.authRoles = "admin"
+        tk.token = Math.random().toString()
+        params.each { k, v ->
+            tk[k] = v
+        }
+        tk.save()
+    }
 }

--- a/rundeckapp/src/test/groovy/rundeck/interceptors/SetUserInterceptorSpec.groovy
+++ b/rundeckapp/src/test/groovy/rundeck/interceptors/SetUserInterceptorSpec.groovy
@@ -2,8 +2,12 @@ package rundeck.interceptors
 
 import com.dtolabs.rundeck.core.authentication.Group
 import com.dtolabs.rundeck.core.authentication.Username
+import com.dtolabs.rundeck.core.authentication.tokens.AuthTokenType
 import grails.core.GrailsApplication
+import grails.testing.gorm.DataTest
 import grails.testing.web.interceptor.InterceptorUnitTest
+import rundeck.AuthToken
+import rundeck.User
 import rundeck.UtilityTagLib
 import rundeck.codecs.HTMLAttributeCodec
 import rundeck.codecs.HTMLContentCodec
@@ -13,10 +17,15 @@ import rundeck.services.UserService
 import spock.lang.Specification
 
 import javax.security.auth.Subject
+import javax.servlet.ServletContext
 import java.security.Principal
 
-class SetUserInterceptorSpec extends Specification implements InterceptorUnitTest<SetUserInterceptor> {
+class SetUserInterceptorSpec extends Specification implements InterceptorUnitTest<SetUserInterceptor>, DataTest {
 
+    void setupSpec() {
+        mockDomain User
+        mockDomain AuthToken
+    }
     def setup() {
     }
 
@@ -93,5 +102,36 @@ class SetUserInterceptorSpec extends Specification implements InterceptorUnitTes
         null         | "auser"  | ["grp1"]         | true
 
 
+    }
+
+    def "lookupToken"() {
+
+        setup:
+        User u1 = new User(login: "admin")
+        User u2 = new User(login: "whk")
+        AuthToken userTk1 = new AuthToken(token: "123",user:u1,authRoles:"admin",type: null)
+        AuthToken userTk2 = new AuthToken(token: "456",user:u1,authRoles:"admin",type: AuthTokenType.USER)
+        AuthToken whkTk = new AuthToken(token: "789",user:u2,authRoles:"admin",type:AuthTokenType.WEBHOOK)
+        u1.save()
+        u2.save()
+        userTk1.save()
+        userTk2.save()
+        whkTk.save()
+        def svCtx = Mock(ServletContext)
+
+        when:
+        String result = interceptor.lookupToken(tk,svCtx,webhookToken)
+
+        then:
+        result == expected
+
+        where:
+        tk|webhookToken|expected
+        "123"|true|null
+        "123"|false|"admin"
+        "456"|true|null
+        "456"|false|"admin"
+        "789"|true|"whk"
+        "789"|false|null
     }
 }

